### PR TITLE
Minecraft image

### PR DESCRIPTION
**Pull request by mrkeith782**
refactors the ItemGenCommand into a few different classes

### CODE REVIEW

File: src/main/java/net/hypixel/nerdbot/generator/StringColorParser.java [hunk 1]
```diff
@@ -0,0 +1,547 @@
...
+                        "string: " + description + "\n" +
+                        "If you see this debug, please ping a developer. Thanks!\n";
+                event.getHook().sendMessage(debug).setEphemeral(true).queue();
+                return null;
+            }
+
+            boolean noColorFlag = false;
+            /* This block checks colors, newline characters, soft-wrapping,
+             * and changes the text depending on those checks.
+             */
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/generator/StringColorParser.java:45]:
```suggestion
            //This block checks colors, newline characters, soft-wrapping,
            //and changes the text depending on those checks.
```

File: src/main/java/net/hypixel/nerdbot/generator/MinecraftImage.java [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+public class MinecraftImage {
+    int locationX = 10;
+    int locationY = 25;
+
+    GraphicsEnvironment ge;
+    Graphics2D g2d;
+    final Font minecraftFont;
+    final Font minecraftBold;
+    MCColor currentColor;
+    boolean boldFlag = false; //True if we're currently printing with bold, false if not.
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/generator/MinecraftImage.java:22]:
```suggestion
    private GraphicsEnvironment ge;
    private Graphics2D g2d;
    private final Font minecraftFont;
    private final Font minecraftBold;
    private MCColor currentColor;
    private boolean boldFlag = false; //True if we're currently printing with bold, false if not.
```

File: src/main/java/net/hypixel/nerdbot/generator/MinecraftImage.java [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+        this.currentColor = defaultColor;
+    }
+
+    public MinecraftImage(int imageWidth, int linesToPrint) {
+        this.ge = GraphicsEnvironment.getLocalGraphicsEnvironment();
+        this.minecraftFont = initFont("/Minecraft/minecraft.ttf", 16f);
+        this.minecraftBold = initFont("/Minecraft/3_Minecraft-Bold.otf", 22f);
+        this.g2d = initG2D(imageWidth, linesToPrint * 23);
+        this.currentColor = MCColor.GRAY;
+    }
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/generator/MinecraftImage.java:39]: Couldn't we use the other constructor and set the third parameter to `MCColor.GRAY`?

File: src/main/java/net/hypixel/nerdbot/generator/MinecraftImage.java [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+                } else {
+                    //We do this to prevent monospace bullshit
+                    subWord.append(line.charAt(colorStartIndex));
+                }
+            }
+
+            drawStringWithBackground(subWord.toString());
+            newLine();
+        }
+
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/generator/MinecraftImage.java:108]:
```suggestion

```

File: src/main/java/net/hypixel/nerdbot/generator/StringColorParser.java [hunk 5]
```diff
@@ -0,0 +1,547 @@
...
+import com.freya02.botcommands.api.application.slash.GuildSlashEvent;
+import net.hypixel.nerdbot.util.skyblock.Gemstone;
+import net.hypixel.nerdbot.util.skyblock.MCColor;
+import net.hypixel.nerdbot.util.skyblock.Stat;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.ArrayList;
+
+public class StringColorParser {
+    /**
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/generator/StringColorParser.java:12]:
```suggestion

    /**
```

**Review by Aerhhh — CHANGES_REQUESTED** [review]

File: src/main/java/net/hypixel/nerdbot/generator/MinecraftImage.java (hunk 3 above)
Comment (mrkeith782) [reply src/main/java/net/hypixel/nerdbot/generator/MinecraftImage.java:39]: Yeah sure that's ez

File: src/main/java/net/hypixel/nerdbot/generator/StringColorParser.java [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+            }
+
+            boolean noColorFlag = false;
+            // This block checks colors, newline characters, soft-wrapping,
+            // and changes the text depending on those checks.
+            if (description.length() != charIndex + 1) {
+                //Color parsing
+                if (description.charAt(charIndex) == '%' && description.charAt(charIndex + 1) == '%') {
+                    int endCharIndex = 0;
+
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/generator/StringColorParser.java:67]:
```suggestion

```

File: src/main/java/net/hypixel/nerdbot/generator/StringColorParser.java [hunk 7]
```diff
@@ -0,0 +1,309 @@
...
+                    //If a parameter can be passed, put that here.
+                    StringBuilder specialSubString = new StringBuilder();
+                    boolean specialSubStringFlag = false;
+                    int specialSubStringIndex = -1;
+                    for (int i = charIndex + 2; i < charIndex + 100; i++) { //get char
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/generator/StringColorParser.java:72]: I don't think this comment is needed

File: src/main/java/net/hypixel/nerdbot/generator/StringColorParser.java [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+                            break;
+                        }
+
+                        if (i + 2 > description.length()) {
+                            endCharIndex = -1;
+                            break;
+                        }
+                    }
+
+                    if (endCharIndex != -1) { //If we can't find the end percents, just continue
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/generator/StringColorParser.java:109]: Move this comment to the line above

File: src/main/java/net/hypixel/nerdbot/generator/StringColorParser.java [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+                        charIndex += 2; //move away from color code
+                        String getSpecialString = description.substring(charIndex, endCharIndex);
+
+                        boolean foundColor = false; //True if we find a valid color, stat, or gemstone.
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/generator/StringColorParser.java:113]: Move this comment to the line above

File: src/main/java/net/hypixel/nerdbot/generator/StringColorParser.java [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+                        }
+
+                        if (i + 2 > description.length()) {
+                            endCharIndex = -1;
+                            break;
+                        }
+                    }
+
+                    if (endCharIndex != -1) { //If we can't find the end percents, just continue
+                        charIndex += 2; //move away from color code
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/generator/StringColorParser.java:110]: Move this comment to the line above

File: src/main/java/net/hypixel/nerdbot/generator/StringColorParser.java [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+
+                //Shorthand Color Parsing
+                if (description.charAt(charIndex) == '&' && description.charAt(charIndex + 1) != ' ') {
+                    for(MCColor color : colors) {
+                        if (color.getColorCode() == description.charAt(charIndex + 1)) {
+                            currString.append("%%").append(color).append("%%");
+                            break;
+                        }
+                    }
+                    if ('l' == description.charAt(charIndex + 1)) {
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/generator/StringColorParser.java:197]:
```suggestion

                    if ('l' == description.charAt(charIndex + 1)) {
```

File: src/main/java/net/hypixel/nerdbot/generator/StringColorParser.java [hunk 12]
```diff
@@ -0,0 +1,309 @@
...
+                    charIndex += 2;
+                    continue;
+                }
+
+                //Softwrap parsing
+                if (description.charAt(charIndex) == ' ') {
+                    charIndex++;
+
+                    int colorCheck = 36; //An extra buffer so we don't wrap colors
+                    boolean newLineFlag = true; //True if we need to make a newline to paste the next word
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/generator/StringColorParser.java:217]: Redundant comment

File: src/main/java/net/hypixel/nerdbot/generator/StringColorParser.java [hunk 13]
```diff
@@ -0,0 +1,309 @@
...
+                    for (int i = charIndex; i < charIndex + (colorCheck - lineLength); i++) {
+                        if (i + 1 > description.length()) {
+                            newLineFlag = false;
+                            break;
+                        }
+                        if (description.charAt(i) == ' ') {
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/generator/StringColorParser.java:223]:
```suggestion

                        if (description.charAt(i) == ' ') {
```

File: src/main/java/net/hypixel/nerdbot/generator/StringColorParser.java [hunk 14]
```diff
@@ -0,0 +1,309 @@
...
+                            newLineFlag = false;
+                            break;
+                        }
+                        if (description.charAt(i) == '%' && description.charAt(i + 1) == '%') {
```
Comment (Aerhhh) [inline src/main/java/net/hypixel/nerdbot/generator/StringColorParser.java:227]:
```suggestion

                        if (description.charAt(i) == '%' && description.charAt(i + 1) == '%') {
```

**Review by Aerhhh — CHANGES_REQUESTED** [review]

**Review by Aerhhh — APPROVED** [review]

**Pull request merged**